### PR TITLE
gRPC

### PR DESCRIPTION
**Pull request by ajax-marchenko-s**

### CODE REVIEW

File: domainservice/src/main/kotlin/ua/marchenko/artauction/auction/controller/nats/AddAuctionNatsController.kt [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+import ua.marchenko.internal.input.reqreply.auction.CreateAuctionRequest as CreateAuctionRequestProto
+
+@Controller
+class AddAuctionNatsController(
+    private val auctionService: AuctionService,
+    override val connection: Connection,
+    private val clock: Clock,
+) : NatsController<CreateAuctionRequestProto, CreateAuctionResponseProto> {
+
+    override val subject: String = NatsSubject.AuctionNatsSubject.CREATE
```
Comment (ajax-torbin-v) [inline domainservice/src/main/kotlin/ua/marchenko/artauction/auction/controller/nats/AddAuctionNatsController.kt:26]:
```suggestion
    override val subject: String = NatsSubject.Auction.CREATE
```

File: domainservice/src/main/kotlin/ua/marchenko/artauction/auction/controller/nats/AddAuctionNatsController.kt [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    override val parser: Parser<CreateAuctionRequestProto> = CreateAuctionRequestProto.parser()
+
+    override fun errorResponse(throwable: Throwable): CreateAuctionResponseProto =
+        throwable.toCreateAuctionFailureResponseProto()
+
+    override fun handle(request: CreateAuctionRequestProto): Mono<CreateAuctionResponseProto> {
+        return auctionService.save(request.toCreateAuctionRequest(clock))
+            .map { it.toCreateAuctionSuccessResponseProto(clock) }
+            .onErrorResume {
+                log.error("Error in CreateAuction", it)
```
Comment (ajax-torbin-v) [inline domainservice/src/main/kotlin/ua/marchenko/artauction/auction/controller/nats/AddAuctionNatsController.kt:39]:
```suggestion
                log.error("Error in CreateAuction for {}",request,  it)
```

File: domainservice/src/test/kotlin/ua/marchenko/artauction/auction/controller/nats/AddAuctionNatsControllerTest.kt [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+    @Test
+    fun `should save new auction and return AuctionResponse with data from CreateAuctionRequest`() {
+        // GIVEN
+        val savedArtist = userRepository.save(MongoUser.random(id = null)).block()
+        val savedArtwork =
+            artworkRepository.save(MongoArtwork.random(artistId = savedArtist!!.id!!.toHexString())).block()
+     val request = AuctionProtoFixture.randomCreateAuctionRequestProto(artworkId = savedArtwork!!.id!!.toHexString())
+        val expectedResponse = CreateAuctionResponseProto.newBuilder().also { builder ->
+            builder.successBuilder.setAuction(
```
Comment (ajax-torbin-v) [inline domainservice/src/test/kotlin/ua/marchenko/artauction/auction/controller/nats/AddAuctionNatsControllerTest.kt:34]:
```suggestion
        val expectedResponse = CreateAuctionResponseProto.newBuilder().apply {
            successBuilder.setAuction(
```

File: domainservice/src/test/kotlin/ua/marchenko/artauction/auction/controller/nats/GetAllAuctionsNatsControllerTest.kt [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+            ).block()!!.toAuctionProto(clock),
+            auctionRepository.save(
+                MongoAuction.random(
+                    id = null,
+                    artworkId = artworks[1].id!!,
+                    startedAt = LocalDateTime.now().truncatedTo(ChronoUnit.MILLIS),
+                    finishedAt = LocalDateTime.now().plusDays(1).truncatedTo(ChronoUnit.MILLIS)
+                )
+            ).block()!!.toAuctionProto(clock),
+        )
```
Comment (ajax-torbin-v) [inline domainservice/src/test/kotlin/ua/marchenko/artauction/auction/controller/nats/GetAllAuctionsNatsControllerTest.kt:54]: If we will want to add more artworks to our list to make it, for example, 5, we will just copy this?

File: domainservice/src/test/kotlin/ua/marchenko/artauction/auction/controller/nats/GetAllAuctionsNatsControllerTest.kt [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        val request = FindAllAuctionsRequestProto.newBuilder().setPage(0).setLimit(100).build()
```
Comment (ajax-torbin-v) [inline domainservice/src/test/kotlin/ua/marchenko/artauction/auction/controller/nats/GetAllAuctionsNatsControllerTest.kt:55]: 🧙‍♂️ 🪄  number

File: domainservice/src/test/kotlin/ua/marchenko/artauction/auction/mapper/AuctionMapperProtoTest.kt [hunk 6]
```diff
@@ -17,16 +17,121 @@ import org.junit.jupiter.api.assertThrows
...
+    fun `should return CreateAuctionRequest from CreateAuctionRequestProto`() {
+        // GIVEN
+        val fixedClock = Clock.fixed(Instant.now(), ZoneId.systemDefault())
+        val requestProto = randomCreateAuctionRequestProto()
+        val expectedResponse = CreateAuctionRequest(
+            requestProto.artworkId,
+            requestProto.startBid.toBigDecimal(),
+            requestProto.startedAt.toLocalDateTime(fixedClock),
+            requestProto.finishedAt.toLocalDateTime(fixedClock),
+        )
```
Comment (ajax-torbin-v) [inline domainservice/src/test/kotlin/ua/marchenko/artauction/auction/mapper/AuctionMapperProtoTest.kt:45]: Why we need to do it like this?

File: domainservice/src/test/kotlin/ua/marchenko/artauction/auction/mapper/AuctionMapperProtoTest.kt [hunk 7]
```diff
@@ -17,16 +17,121 @@ import org.junit.jupiter.api.assertThrows
...
+        assertEquals(expectedResponse, result)
+    }
+
+    @Test
+    fun `should build CreateAuctionResponseProto success from MongoAuction`() {
+        // GIVEN
+        val fixedClock = Clock.fixed(Instant.now(), ZoneId.systemDefault())
+        val mongoAuction = MongoAuction.random()
+        val expectedResponse = CreateAuctionResponseProto.newBuilder().also {
+            it.successBuilder.auction = mongoAuction.toAuctionProto(fixedClock)
```
Comment (ajax-torbin-v) [inline domainservice/src/test/kotlin/ua/marchenko/artauction/auction/mapper/AuctionMapperProtoTest.kt:60]:
```suggestion
        val expectedResponse = CreateAuctionResponseProto.newBuilder().apply {
            successBuilder.auction = mongoAuction.toAuctionProto(fixedClock)
```

File: gateway/src/main/kotlin/ua/marchenko/gateway/auction/mapper/AuctionProtoMapperExtension.kt [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+
+        FindAuctionByIdResponseProtoInternal.ResponseCase.FAILURE -> {
+            when (failure.errorCase!!) {
+                FindAuctionByIdResponseProtoInternal.Failure.ErrorCase.NOT_FOUND_BY_ID ->
+                    throw AuctionNotFoundException(message = failure.message)
+
+                FindAuctionByIdResponseProtoInternal.Failure.ErrorCase.ERROR_NOT_SET ->
+                    error(failure.message)
+            }
+        }
```
Comment (ajax-torbin-v) [inline gateway/src/main/kotlin/ua/marchenko/gateway/auction/mapper/AuctionProtoMapperExtension.kt:34]: Extract this

File: gateway/src/main/kotlin/ua/marchenko/gateway/auction/mapper/AuctionProtoMapperExtension.kt [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+
+        CreateAuctionResponseProtoInternal.ResponseCase.RESPONSE_NOT_SET -> error("Response not set")
+
+        CreateAuctionResponseProtoInternal.ResponseCase.FAILURE -> {
+            when (failure.errorCase!!) {
+                CreateAuctionResponseProtoInternal.Failure.ErrorCase.INVALID_AUCTION_OPERATION ->
+                    throw InvalidAuctionOperationException(failure.message)
+
+                CreateAuctionResponseProtoInternal.Failure.ErrorCase.ERROR_NOT_SET ->
+                    error(failure.message)
```
Comment (ajax-torbin-v) [inline gateway/src/main/kotlin/ua/marchenko/gateway/auction/mapper/AuctionProtoMapperExtension.kt:60]: failure.message.orEmpty() ?

File: grpc-api/src/main/proto/ua/marchenko/grpcapi/input/reqreply/auction/create_auction.proto [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+import "ua/marchenko/commonmodels/error.proto";
+import "ua/marchenko/commonmodels/auction/auction.proto";
+import "ua/marchenko/commonmodels/general/big_decimal.proto";
+import "google/protobuf/timestamp.proto";
+
+option java_multiple_files = true;
+
+message CreateAuctionRequest {
+  string artwork_id = 1;
+  ua.marchenko.commonmodels.general.BigDecimal start_bid = 2;
```
Comment (ajax-torbin-v) [inline grpc-api/src/main/proto/ua/marchenko/grpcapi/input/reqreply/auction/create_auction.proto:14]: Remove full package name qualifying

File: gateway/src/main/kotlin/ua/marchenko/gateway/auction/mapper/AuctionProtoMapperExtension.kt [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+            when (failure.errorCase!!) {
+                FindAuctionByIdResponseProtoInternal.Failure.ErrorCase.NOT_FOUND_BY_ID ->
+                    throw AuctionNotFoundException(message = failure.message)
+
+                FindAuctionByIdResponseProtoInternal.Failure.ErrorCase.ERROR_NOT_SET ->
+                    error(failure.message)
+            }
+        }
+    }
+}
```
Comment (ajax-yurkovskyi-t) [inline gateway/src/main/kotlin/ua/marchenko/gateway/auction/mapper/AuctionProtoMapperExtension.kt:36]: u can decompose this

**Review by ajax-yurkovskyi-t — COMMENTED** [review]
Good job

File: domainservice/src/test/kotlin/ua/marchenko/artauction/auction/controller/nats/GetAllAuctionsNatsControllerTest.kt (hunk 4 above)
Comment (ajax-marchenko-s) [reply domainservice/src/test/kotlin/ua/marchenko/artauction/auction/controller/nats/GetAllAuctionsNatsControllerTest.kt:54]: there are different artworkId, so I cant just write `List (2) {auctionRepository.save(MongoAuction.random...)}`

File: gateway/src/main/kotlin/ua/marchenko/gateway/common/nats/NatsClient.kt [hunk 12]
```diff
@@ -16,4 +26,20 @@ class NatsClient(private val natsConnection: Connection) {
         return Mono.fromFuture { natsConnection.request(subject, payload.toByteArray()) }
             .map { parser.parseFrom(it.data) }
     }
+
+    fun subscribeToCreatedAuction(): Flux<AuctionProto> {
+        val sink = Sinks.many().unicast().onBackpressureBuffer<AuctionProto>()
+        val subscription = dispatcher.subscribe(NatsSubject.AuctionNatsSubject.CREATED_EVENT) { message ->
+            sink.tryEmitNext(AuctionProto.parseFrom(message.data))
+        }
```
[the remark this answers is not in the document]
Comment (ajax-marchenko-s) [reply gateway/src/main/kotlin/ua/marchenko/gateway/common/nats/NatsClient.kt]: add runCatching and log errors

File: domainservice/src/test/kotlin/ua/marchenko/artauction/auction/controller/nats/GetAuctionByIdNatsControllerTest.kt [hunk 13]
```diff
@@ -0,0 +1,59 @@
+package ua.marchenko.artauction.auction.controller.nats
+
+import artwork.random
+import auction.random
```
Comment (ajax-shulepa-a) [inline domainservice/src/test/kotlin/ua/marchenko/artauction/auction/controller/nats/GetAuctionByIdNatsControllerTest.kt:4]:
```suggestion
import ua.marchenko.artauction.artwork.random
import ua.marchenko.artauction.auction.random
```

**Review by ajax-shulepa-a — APPROVED** [review]

File: domainservice/src/test/kotlin/ua/marchenko/artauction/auction/controller/nats/GetAuctionByIdNatsControllerTest.kt (hunk 13 above)
Comment (ajax-marchenko-s) [reply domainservice/src/test/kotlin/ua/marchenko/artauction/auction/controller/nats/GetAuctionByIdNatsControllerTest.kt:4]: Random method is in the testFixtures `package auction`, but `ua.marchenko.artauction.auction` is the main package, so I can't change it (without changing package in testFixtures and imports in my all test)

Comment (ajax-marchenko-s) [reply domainservice/src/test/kotlin/ua/marchenko/artauction/auction/controller/nats/GetAuctionByIdNatsControllerTest.kt:4]: done in new PR

**Pull request merged**